operator/validator/metadata/syncer.go

```diff
@@ -0,0 +1,288 @@
+	select {
+	case <-ctx.Done():
+		return ValidatorUpdate{}, false, ctx.Err()
+	default:
+	}
+
+	shares := u.sharesBatchForUpdate(ctx)
+	if len(shares) == 0 {
+		return ValidatorUpdate{}, false, nil
+	}
+
+	pubKeys := make([]spectypes.ValidatorPK, len(shares))
+	for i, s := range shares {
+		pubKeys[i] = s.ValidatorPubKey
+	}
+
+	indicesBefore := u.allActiveIndices(ctx, u.beaconNetwork.GetBeaconNetwork().EstimatedCurrentEpoch())
+
+	validators, err := u.Sync(ctx, pubKeys)
+	if err != nil {
+		return ValidatorUpdate{}, false, fmt.Errorf("update metadata: %w", err)
+	}
+
+	indicesAfter := u.allActiveIndices(ctx, u.beaconNetwork.GetBeaconNetwork().EstimatedCurrentEpoch())
+
+	update := ValidatorUpdate{
+		IndicesBefore: indicesBefore,
+		IndicesAfter:  indicesAfter,
+		Validators:    validators,
+	}
+
+	return update, len(shares) < batchSize, nil
+}
+
+// sharesBatchForUpdate returns non-liquidated shares from DB that are most deserving of an update, it relies on share.Metadata.lastUpdated to be updated in order to keep iterating forward.
+func (u *ValidatorSyncer) sharesBatchForUpdate(_ context.Context) []*ssvtypes.SSVShare {
+	// TODO: use context, return if it's done
+	var staleShares, newShares []*ssvtypes.SSVShare
+	u.shareStorage.Range(nil, func(share *ssvtypes.SSVShare) bool {
+		if share.Liquidated {
+			return true
+		}
+
+		shareMetadataStaleAfter := u.syncInterval
+		// also check last metadata update to differentiate new shares from stale ones
+		if !share.HasBeaconMetadata() && share.MetadataLastUpdated().IsZero() {
+			newShares = append(newShares, share)
+		} else if time.Since(share.MetadataLastUpdated()) > shareMetadataStaleAfter {
+			staleShares = append(staleShares, share)
+		}
+		return len(newShares) < batchSize
+	})
+
+	// Combine validators up to batchSize, prioritizing the new ones.
+	shares := newShares
+	if remainder := batchSize - len(shares); remainder > 0 {
+		end := remainder
+		if end > len(staleShares) {
+			end = len(staleShares)
+		}
+		shares = append(shares, staleShares[:end]...)
+	}
+
+	for _, share := range shares {
+		share.SetMetadataLastUpdated(time.Now())
+	}
+	return shares
+}
+
+// TODO: Create a wrapper for share storage that contains all common methods like AllActiveIndices and use the wrapper.
+func (u *ValidatorSyncer) allActiveIndices(_ context.Context, epoch phase0.Epoch) []phase0.ValidatorIndex {
+	var indices []phase0.ValidatorIndex
+
+	// TODO: use context, return if it's done
+	u.shareStorage.Range(nil, func(share *ssvtypes.SSVShare) bool {
+		if share.IsParticipating(epoch) {
+			indices = append(indices, share.BeaconMetadata.Index)
+		}
+		return true
+	})
+
+	return indices
+}
+
+func (u *ValidatorSyncer) sleep(ctx context.Context, d time.Duration) (slept bool) {
+	timer := time.NewTimer(d)
+	defer timer.Stop()
+
+	select {
+	case <-ctx.Done():
+		return false
+	case <-timer.C:
+		return true
+	}
+}
```

Comment:
why don't you use wrap ctx with deadline/timeout instead of passing in duration?

Reply:
@y0sher If the context is canceled, I break the loop, and if the timer fires I continue the loop, so if I wrap it to a context, then I won't know if I still need to continue iterating. Without this logic, I'd likely get rid of this method